optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/tri/TriConstraintCollector.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.api.score.stream.tri;
+
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+import org.optaplanner.core.api.function.QuadFunction;
+import org.optaplanner.core.api.function.TriFunction;
+import org.optaplanner.core.api.score.stream.ConstraintCollectors;
+import org.optaplanner.core.api.score.stream.ConstraintStream;
+import org.optaplanner.core.api.score.stream.bi.BiConstraintStream;
+
+/**
+ * Usually created with {@link ConstraintCollectors}.
+ * Used by {@link TriConstraintStream#groupBy(TriFunction, TriConstraintCollector)}, ...
+ * <p>
+ * Loosely based on JDK's {@link Collector}, but it returns an undo operation for each accumulation
+ * to enable incremental score calculation in {@link ConstraintStream constraint streams}.
+ * @param <A> the type of the first matched fact in the source {@link BiConstraintStream}
+ * @param <B> the type of the second matched fact in the source {@link BiConstraintStream}
+ * @param <C> the type of the third matched fact in the source {@link BiConstraintStream}
+ * @param <ResultContainer_> the mutable accumulation type (often hidden as an implementation detail)
+ * @param <Result_> the type of a match fact in the destination {@link ConstraintStream}
+ * @see ConstraintCollectors
+ */
+public interface TriConstraintCollector<A, B, C, ResultContainer_, Result_> {
```

Comment:
I don't mind this being here, good to merge, but this is unrelated to quad steams, right?

Reply:
Well spotted, I'll take it out. Let's make this PR clean.